enhancements/microshift/low-latency-workloads-on-microshift.md

```diff
@@ -0,0 +1,612 @@
+
+Optionally, a new subcommand `microshift doctor low-latency` might be added to main
+MicroShift binary to provide some verification checks if system configuration is matching
+expectations according to our knowledge. It shall not configure system - only report potential problems.
+
+
+### Workflow Description
+
+Workflow consists of two parts:
+1. system and MicroShift configuration
+1. preparing Kubernetes manifests for low latency
+
+#### System and MicroShift configuration
+
+##### OSTree
+
+1. User creates an osbuild blueprint:
+   - (optional) User configures `[customizations.kernel]` in the blueprint if the values are known
+     beforehand. This could prevent from necessary reboot after applying tuned profile.
+   - User adds `kernel-rt` package to the blueprint
+   - User adds `microshift-low-latency` RPM to the blueprint
+   - User enables `microshift-tuned.service`
+   - User supplies additional configs using blueprint:
+     - /etc/tuned/microshift-low-latency-variables.conf
+     - /etc/microshift/config.yaml to configure Kubelet
+1. User builds the blueprint
+1. User deploys the commit / installs the system.
+1. System boots
+1. `microshift-tuned.service` starts (after `tuned.service`, before `microshift.service`):
+   - Saves current kernel args
+   - Applies tuned `microshift-low-latency` profile
+   - Verifies expected kernel args
+     - ostree: `rpm-ostree kargs`
+     - rpm: `grubby`
+   - If the current and expected kernel args are different, reboot the node
+1. Host boots again, everything for low latency is in place,
+   `microshift.service` can continue start up.
+
+##### RPM
+
+1. User installs `microshift-low-latency` RPM.
+1. User creates following configs:
+   - /etc/tuned/microshift-low-latency-variables.conf
+   - /etc/microshift/config.yaml to configure Kubelet
+1. User enables `microshift-tuned.service` or uses `tuned-adm` directly to activate the profile
+   (and reboot the host if needed).
+1. If host was not rebooted, CRI-O and MicroShift services need to be restarted to make new settings
+   active.
+
+#### Preparing low latency workload
+
+- Setting `.spec.runtimeClassName: microshift-low-latency` in Pod spec.
+- Setting Pod's memory limit and memory request to the same value, and
+  setting CPU limit and CPU request to the same value to ensure Pod has guaranteed QoS class.
+- Use annotations to get desired behavior:
+  - `cpu-load-balancing.crio.io: "disable"` - disable CPU load balancing for Pod 
+    (only use with CPU Manager `static` policy and for Guaranteed QoS Pods using whole CPUs)
+  - `cpu-quota.crio.io: "disable"` - disable Completely Fair Scheduler (CFS)
+  - `irq-load-balancing.crio.io: "disable"` - disable interrupt processing
+    (only use with CPU Manager `static` policy and for Guaranteed QoS Pods using whole CPUs)
+  - `cpu-c-states.crio.io: "disable"` - disable C-states
+  - `cpu-freq-governor.crio.io: "<governor>"` - specify governor type for CPU Freq scaling (e.g. `performance`)
```

Comment:
Can we get a link to the available values?

Reply:
I'll track them down in OpenShift docs. Sneak peak: https://docs.okd.io/4.15/scalability_and_performance/low_latency_tuning/cnf-provisioning-low-latency-workloads.html#cnf-configuring-high-priority-workload-pods_cnf-provisioning-low-latency

As for governors: https://www.kernel.org/doc/Documentation/cpu-freq/governors.txt